plugins/out_s3/s3.c

```diff
@@ -1393,91 +1594,140 @@ static void cb_s3_flush(const void *data, size_t bytes,
                       ctx->timer_ms / 1000);
 
         sched = flb_sched_ctx_get();
-        ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
-                                        ctx->timer_ms,
-                                        cb_s3_upload,
-                                        ctx);
+
+        if (ctx->preserve_data_ordering) {
+            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
+                                            ctx->timer_ms, s3_upload_queue, ctx);
+        }
+        else {
+            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
+                                            ctx->timer_ms, cb_s3_upload, ctx);
+        }
         if (ret == -1) {
             flb_plg_error(ctx->ins, "Failed to create upload timer");
             FLB_OUTPUT_RETURN(FLB_RETRY);
         }
         ctx->timer_created = FLB_TRUE;
     }
 
-    json = flb_pack_msgpack_to_json_format(data, bytes,
-                                           FLB_PACK_JSON_FORMAT_LINES,
-                                           ctx->json_date_format,
-                                           ctx->date_key);
+    chunk = flb_pack_msgpack_to_json_format(data, bytes,
+                                            FLB_PACK_JSON_FORMAT_LINES,
+                                            ctx->json_date_format,
+                                            ctx->date_key);
 
-    if (json == NULL) {
-        flb_plg_error(ctx->ins, "Could not marshal msgpack to JSON");
+    if (chunk == NULL) {
+        flb_plg_error(ctx->ins, "Could not marshal msgpack to output string");
         FLB_OUTPUT_RETURN(FLB_ERROR);
     }
 
-    len = flb_sds_len(json);
+    chunk_size = flb_sds_len(chunk);
 
-    /* Get a chunk candidate matching the given 'Tag' */
-    chunk = s3_store_file_get(ctx, tag, tag_len);
+    /* Get a chunk candidate matching the given 'tag' */
+    upload_file = s3_store_file_get(ctx, tag, tag_len);
 
-    if (chunk != NULL && chunk->failures >= MAX_UPLOAD_ERRORS) {
-        flb_plg_warn(ctx->ins, "Chunk for tag %s failed to send %d times, "
-                     "will not retry", tag, MAX_UPLOAD_ERRORS);
-        s3_store_file_inactive(ctx, chunk);
-        chunk = NULL;
-    }
+    /* Specific to unit tests, will not get called normally */
+    if (s3_plugin_under_test() == FLB_TRUE) {
```

Comment:
better to have a separate func for this unit test logic and only have a func call there to make the main thread logic clear